fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java

```diff
@@ -2591,51 +2602,115 @@ private Bundle createSearchBundle(List<Resource> resources, FHIRSearchContext se
-                        issues.add(FHIRUtil.buildOperationOutcomeIssue(IssueSeverity.WARNING, IssueType.NOT_SUPPORTED, msg, node.path()));
+            }
+
+            // Loop through the resources, looking for versioned references
+            for (Resource resource : resources) {
+                FHIRPathEvaluator evaluator = FHIRPathEvaluator.evaluator();
+                EvaluationContext evaluationContext = new EvaluationContext(resource);
+                for (QueryParameter queryParameter : queryParameters) {
+                    SearchParameter searchParameter = searchParameterMap.get(queryParameter);
+                    if (searchParameter == null) {
+                        searchParameter = SearchUtil.getSearchParameter(resource.getClass(), queryParameter.getCode());
+                    }
+                    Collection<FHIRPathNode> nodes = evaluator.evaluate(evaluationContext, searchParameter.getExpression().getValue());
+                    for (FHIRPathNode node : nodes) {
+                        Reference reference = node.asElementNode().element().as(Reference.class);
+                        ReferenceValue rv = ReferenceUtil.createReferenceValueFrom(reference, ReferenceUtil.getBaseUrl(null));
+                        if (rv.getVersion() != null &&
+                                (rv.getTargetResourceType() == null || rv.getTargetResourceType().equals(queryParameter.getModifierResourceTypeName()))) {
+                            String msg = "Resource with id '" + resource.getId() +
+                                    "' contains a versioned reference in an element used for chained search, but chained search does not act on versioned references.";
+                            issues.add(FHIRUtil.buildOperationOutcomeIssue(IssueSeverity.WARNING, IssueType.NOT_SUPPORTED, msg, node.path()));
+                        }
                     }
                 }
-            } catch (java.lang.UnsupportedOperationException | FHIRPathException uoe) {
-                log.warning(String.format("Search Parameter includes an unsupported operation or bad expression : [%s] [%s] [%s]",
-                    searchParameter.getCode().getValue(), searchParameter.getExpression().getValue(), uoe.getMessage()));
             }
         }
 
         return issues;
     }
 
+    /**
+     * For reference searches against logical ID only, check for existence of multiple resource types
+     * containing the same logical ID in 'match' resource references.
+     *
+     * @param resourceType
+     *            The search resource type.
+     * @param queryParameters
+     *            The list of reference query parameters that only specified a logical ID.
+     * @param resources
+     *            The list of resources to check for multiple resource types.
+     * @throws Exception
+     */
+    private void checkForMatchLogicalIdDuplicates(String resourceType, List<String> queryParameters, List<Resource> resources) throws Exception {
```

Comment:
ugh, I really wish this were simpler; I'm just not sure how to do it without just saying "screw it" :-/

Reply:
Yeah, I just don't think there's an easy way to do it. Good thing is that it's only going to get executed for searches in which there's a logical ID-only reference value specified.

Reply:
It feels like a lot of processing over resources multiple times. 

Could we add a flag to turn this feature off or on?  specifically on line 2670, 2616
it feels like something we may want to have as lenient or strict.

Reply:
I'll make a couple points and then we can decide how to proceed:
1. I've combined checkForMatchLogicalIdDuplicates() and checkForVersionedReference() into one method: performSearchReferenceChecks(), so now we'll only loop through 'match' resources, and only a maximum of one time
2. We'll only do this processing on a chain search or on a search of a reference field where the search parameter value is a logical ID-only, the former being uncommon and the latter being rare.

Given the above, I think a flag would probably be overkill, particularly given the fact that we're doing this processing to be compliant with the FHIR spec.